File: apps/web/src/application/offline-sync-queue.ts
import {
  syncQueueItemSchema,
  syncQueueProcessInputSchema,
  syncQueueProcessResultSchema,
  type NutritionLog,
  type SyncQueueItem,
  type SyncQueueProcessInput,
  type SyncQueueProcessResult,
  type TrainingPlan,
  type WorkoutSessionInput
} from "@flux/contracts";

export interface OfflineQueueStore {
  add(item: SyncQueueItem): Promise<void>;
  list(userId: string): Promise<SyncQueueItem[]>;
  remove(ids: string[]): Promise<void>;
}

export type OfflineSyncIdempotencyMetadata = {
  key: string;
  replayed: boolean;
  ttlSeconds: number;
};

export type OfflineSyncProcessOutcome = SyncQueueProcessResult & {
  idempotency: OfflineSyncIdempotencyMetadata | null;
};

export interface OfflineSyncGateway {
  process(input: SyncQueueProcessInput): Promise<
    SyncQueueProcessResult & { idempotency?: unknown }
  >;
}

export class OfflineSyncQueueUseCase {
  constructor(
    private readonly store: OfflineQueueStore,
    private readonly gateway: OfflineSyncGateway,
    private readonly now: () => Date = () => new Date(),
    private readonly idFactory: () => string = () => crypto.randomUUID()
  ) {}

  async queueTrainingPlan(userId: string, payload: Omit<TrainingPlan, "createdAt">): Promise<void> {
    const item = syncQueueItemSchema.parse({
      id: this.idFactory(),
      userId,
      enqueuedAt: this.now().toISOString(),
      action: {
        type: "create_training_plan",
        payload
      }
    });
    await this.store.add(item);
  }

  async queueWorkoutSession(userId: string, payload: WorkoutSessionInput): Promise<void> {
    const item = syncQueueItemSchema.parse({
      id: this.idFactory(),
      userId,
      enqueuedAt: this.now().toISOString(),
      action: {
        type: "create_workout_session",
        payload
      }
    });
    await this.store.add(item);
  }

  async queueNutritionLog(userId: string, payload: NutritionLog): Promise<void> {
    const item = syncQueueItemSchema.parse({
      id: this.idFactory(),
      userId,
      enqueuedAt: this.now().toISOString(),
      action: {
        type: "create_nutrition_log",
        payload
      }
    });
    await this.store.add(item);
  }

  async listPending(userId: string): Promise<SyncQueueItem[]> {
    return this.store.list(userId);
  }

  async syncPending(userId: string): Promise<OfflineSyncProcessOutcome> {
    const items = await this.store.list(userId);
    const input = syncQueueProcessInputSchema.parse({ userId, items });
    if (input.items.length === 0) {
      return {
        ...syncQueueProcessResultSchema.parse({ acceptedIds: [], rejected: [] }),
        idempotency: null
      };
    }

    const gatewayResponse = await this.gateway.process(input);
    const result = syncQueueProcessResultSchema.parse(gatewayResponse);
    const idempotency = parseOfflineSyncIdempotencyMetadata(gatewayResponse.idempotency);

    if (result.acceptedIds.length > 0) {
      await this.store.remove(result.acceptedIds);
    }

    return {
      ...result,
      idempotency
    };
  }
}

function parseOfflineSyncIdempotencyMetadata(
  input: unknown
): OfflineSyncIdempotencyMetadata | null {
  if (typeof input !== "object" || input === null) {
    return null;
  }

  const candidate = input as {
    key?: unknown;
    replayed?: unknown;
    ttlSeconds?: unknown;
  };

  if (
    typeof candidate.key !== "string" ||
    candidate.key.length === 0 ||
    typeof candidate.replayed !== "boolean" ||
    typeof candidate.ttlSeconds !== "number" ||
    Number.isFinite(candidate.ttlSeconds) === false
  ) {
    return null;
  }

  return {
    key: candidate.key,
    replayed: candidate.replayed,
    ttlSeconds: candidate.ttlSeconds
  };
}

File: apps/ios/Sources/Domain/OfflineSyncQueue.swift
import Foundation

public struct QueuedTrainingPlanInput: Sendable, Equatable {
  public let id: String
  public let userID: String
  public let name: String
  public let weeks: Int
  public let days: [TrainingPlanDay]

  public init(id: String, userID: String, name: String, weeks: Int, days: [TrainingPlanDay]) {
    self.id = id
    self.userID = userID
    self.name = name
    self.weeks = weeks
    self.days = days
  }
}

public enum OfflineSyncAction: Sendable, Equatable {
  case createTrainingPlan(QueuedTrainingPlanInput)
  case createWorkoutSession(CreateWorkoutSessionInput)
  case createNutritionLog(NutritionLog)
}

public struct OfflineSyncQueueItem: Sendable, Equatable, Identifiable {
  public let id: String
  public let userID: String
  public let enqueuedAt: Date
  public let action: OfflineSyncAction

  public init(id: String, userID: String, enqueuedAt: Date, action: OfflineSyncAction) {
    self.id = id
    self.userID = userID
    self.enqueuedAt = enqueuedAt
    self.action = action
  }
}

public struct OfflineSyncRejectedItem: Sendable, Equatable {
  public let id: String
  public let reason: String

  public init(id: String, reason: String) {
    self.id = id
    self.reason = reason
  }
}

public struct OfflineSyncResult: Sendable, Equatable {
  public let acceptedIDs: [String]
  public let rejected: [OfflineSyncRejectedItem]
  public let idempotency: OfflineSyncIdempotencyMetadata?

  public init(
    acceptedIDs: [String],
    rejected: [OfflineSyncRejectedItem],
    idempotency: OfflineSyncIdempotencyMetadata? = nil
  ) {
    self.acceptedIDs = acceptedIDs
    self.rejected = rejected
    self.idempotency = idempotency
  }
}

public struct OfflineSyncIdempotencyMetadata: Sendable, Equatable {
  public let key: String
  public let replayed: Bool
  public let ttlSeconds: Int

  public init(key: String, replayed: Bool, ttlSeconds: Int) {
    self.key = key
    self.replayed = replayed
    self.ttlSeconds = ttlSeconds
  }
}

public protocol OfflineSyncQueueRepository: Sendable {
  func enqueue(item: OfflineSyncQueueItem) async throws
  func list(userID: String) async throws -> [OfflineSyncQueueItem]
  func remove(ids: [String]) async throws
}

public protocol OfflineSyncGateway: Sendable {
  func process(userID: String, items: [OfflineSyncQueueItem]) async throws -> OfflineSyncResult
}
